packages/medusa/src/services/discount-condition.ts

```diff
@@ -145,6 +147,31 @@ class DiscountConditionService extends BaseService {
           this.discountRepository_
         )
 
+        let discount
+        if (typeof discountIdOrDiscount === `string`) {
```

Comment:
@pKorsholm have added this guard to `upsertCondition` to ensure, that we throw on non-existing discounts. It also allows you to provide a Discount object to the method instead of an id.

Reply:
nit: Instead of including a guard here, I think we should move the retrieve to the API since in both `create-condition` and `update-condition` we have a discount service and can use the standard `retrieve`. In the discountService where this method is also invoked from, we in both cases have the discount fetched.

I think we can remove the branch where we fetch the discount and use just a discount directly since this is a duplication of `discountService.retrieve`. Removing the option to pass a string wouldn't change anything since you could fetch the discount in the API and pass it along as a parameter instead, the database operations, in that case, would be the same, however, it would remove the use of `discountRepository.findOne`.

Reply:
one approach that I used to do was to have some middlewares that are retrieving the resources. For example:

`/discounts/:id/conditions/:cid/test`
It would include two middlewares `fetchDiscount` and `fetchDiscountCondition` that both do what there names mean and attach the resource on the req. It also allow to check for example if the discount condition belongs to the discount, or if the loggedInUser has the right to do the action he is trying to do and so on. It also centralise some basic fetch that can them compose the API as well as simplify the constraints checking and controller logic.

In the end, the discount and discountCondition (in my example) are both accessible from the controller through `req` and at that point we already know that the condition belongs to the discount and that the user has the right to update it for example.

This is something we can talk about to introduce that approach that works pretty well in many projects.